### ISSUE TITLE
V and Q Loss increasing when training with IQL

### ISSUE THREAD

**Issue by shinkuan** [Cryolite/kanachan#45]
The loss keeps increasing. What config I may have set wrong?

Is the LR too high? I was using default value.

Ran the training using: 
`torchrun --nproc_per_node gpu --standalone -m kanachan.training.iql.train training_data=/workspace/data/annotate4rl_00000.txt num_workers=2 device=cuda encoder=bert_base decoder=double reward_plugin=/workspace/kanachan/kanachan/training/iql/get_reward.py discount_factor=1.0 target_update_rate=0.1 checkpointing=true batch_size=200 snapshot_interval=3000000 expectile=0.9`

Reward function: (Pseudo Code)
```
reward = 0
reward += (6-shanten) * 2

if action is discardTile:
    if (じゅんめ) > 6:
        safeTileReward = evaluateSafeTile()
        reward += safeTileReward 

if action is Ron or Zimo:
    reward += 10

if game_rank == 1:
    reward += 10
if game_rank == 2:
    reward += 3
if game_rank == 3:
    reward -= 3 
if game_rank == 4:
    reward -= 10

return reward
```

![VN](https://github.com/Cryolite/kanachan/assets/35415788/2396c639-274a-4ad0-916f-77a965b88150)
![VL](https://github.com/Cryolite/kanachan/assets/35415788/56b9e538-289b-4870-99ce-90fd98888ab9)

![Q1N](https://github.com/Cryolite/kanachan/assets/35415788/b74c5cf4-63b2-4c2c-a84a-fb0ec5e7b512)
![Q1L](https://github.com/Cryolite/kanachan/assets/35415788/5306426c-1215-48bf-a7eb-ccba1d7ac91b)

![Q2N](https://github.com/Cryolite/kanachan/assets/35415788/10edace1-6401-49ce-86a7-a4de555b5452)
9f8)
![Q2L](https://github.com/Cryolite/kanachan/assets/35415788/fd551a6a-9852-4922-8c38-4a909c8ee584)

[train.log](https://github.com/Cryolite/kanachan/files/12175574/train_copy.txt)

**Comment by KimamanaNeko** [comment]
The goal of model training is to minimize the value of the loss function, thus a lower q1 loss or q2 loss value indicates that the model's predictions are closer to the actual outcomes, and therefore, the model is performing better.
I think your reward signal are improperly set, this could cause the model's behavior to deviate from what is expected, leading to an increase in q loss.

**Comment by shinkuan** [comment]
> thus a lower q1 loss or q2 loss value indicates that the model's predictions are closer to the actual outcomes

What about V loss, what does V loss represent?

> I think your reward signal are improperly set, this could cause the model's behavior to deviate from what is expected, leading to an increase in q loss.

The reward function I gave tend to give more reward when close to the win (shanten) of the game.
I don't know whether I should give reward base on `the decision it made is good` or base on `how close it is right now to winning the game.`

I tried to lower optimizer's learning rate. And that seems to solve the problem:
![Q1L](https://github.com/Cryolite/kanachan/assets/35415788/916f400a-4bc7-44fc-9cdc-d916ddf7eb50)
![Q1N](https://github.com/Cryolite/kanachan/assets/35415788/570be16e-5d1a-48b7-a0db-a89dd57601c5)
![Q2L](https://github.com/Cryolite/kanachan/assets/35415788/6108a9b7-7f11-457f-8b7c-56f479d9bfc0)
![Q2N](https://github.com/Cryolite/kanachan/assets/35415788/7172b864-f820-4338-aaeb-84418276d7b3)
![VL](https://github.com/Cryolite/kanachan/assets/35415788/31018e56-0aa5-48b1-bd68-a9632e25657f)
![VN](https://github.com/Cryolite/kanachan/assets/35415788/5cfa2b31-b51e-4b92-b763-25ed0676e2d9)

**Comment by KimamanaNeko** [comment]
v loss and q loss refer to the losses of the value function and the action-value function respectively.
v loss reflects the accuracy of the model in evaluating state values
Therefore, it should be as low as possible
